File: src/java/examples/pingpong/Ping.java
/*
* Copyright 2008, 2009 Complex Automata Simulation Technique (COAST) consortium
* Copyright 2010-2013 Multiscale Applications on European e-Infrastructures (MAPPER) project
*
* GNU Lesser General Public License
* 
* This file is part of MUSCLE (Multiscale Coupling Library and Environment).
* 
* MUSCLE is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* MUSCLE is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
* 
* You should have received a copy of the GNU Lesser General Public License
* along with MUSCLE.  If not, see <http://www.gnu.org/licenses/>.
*/
package examples.pingpong;

import muscle.core.ConduitEntrance;
import muscle.core.ConduitExit;
import muscle.core.kernel.CAController;

public class Ping extends CAController {
	private ConduitEntrance<byte[]> entrance;
	private ConduitExit<byte[]> exit;

	@Override
	protected void addPortals() {
		entrance = addEntrance("out", byte[].class);
		exit = addExit("in", byte[].class);
	}

	@Override
	protected void execute() {
		// How many steps for a single test
		int steps = getIntProperty("steps");

		// How many steps total will be done
		int max_timesteps = getIntProperty("max_timesteps") + 1;
		
		// How many steps total will be done
		int runs = getIntProperty("same_size_runs");
		
		// size in B
		int size = getIntProperty("start_kiB_per_message")*1024;

		// helper with results for a single test
		long[] totalTimes = new long[runs];
		byte[] data = new byte[1024];

		// Making noise in order to give time for JVM to stabilize
		System.out.print("Preparing");
		
		int prepare_steps = getIntProperty("preparation_steps");
		
		for (int i = 0; i < prepare_steps; i++) {
			entrance.send(data);
			exit.receive();
			if (i % 5 == 0) {
				System.out.print(".");
			}
		}
		System.out.println("\n\nValues are NOT divided by 2. Each value is calculated for RTT.");
		System.out.println("Sending " + max_timesteps + " messages in total. For each data size, " + runs + " tests are performed, each sending " + steps + " messages.\n");
		
		System.out.printf("|=%10s|=%10s|=%10s|=%10s|=%10s|=%13s|\n","Size (kiB)","Total (ms)","Avg (us)","StdDev(ms)","StdDev(%)","Speed (MiB/s)");
		
		for (int i = 0; i < getIntProperty("tests_count"); i++) {
			long sum = doComputation(size, totalTimes, runs, steps);
			
			printOutcomes(totalTimes, sum, size, steps);
			if (size == 0) {
				size = 1024;
			} else if (size < Integer.MAX_VALUE / 2) {
				size *= 2;
			} else {
				break;
			}
		}
	}
	
	private long doComputation(int size, long[] totalTimes, int runs, int steps) {
		byte[] data = new byte[size];
		long sum = 0;
		for (int test = 0; test < runs; test++) {
			long tAll = System.nanoTime();
			for (int i = 0; i < steps; i++) {
				if (willStop()) {
					System.out.println("willStop now!");
					return 0;
				}
				// The step
				entrance.send(data);
				exit.receive();
			}
			sum += totalTimes[test] = System.nanoTime() - tAll;
		}
		
		return sum;
	}
	
	private void printOutcomes(long[] totalTimes, long sum, int size, int steps) {	
		double avg = sum/((double)(totalTimes.length * steps));
		double stdDev = stdDev(totalTimes, avg, steps);

		// nano -> micro
		avg /= 1000d;
		stdDev /= 1000d;
		// MB/s -> 2*((bytes/1024*1024)/(micro/1000000)) -> 2*1000000*Bytes/(1024*1024*micro) ->  15625*size / 8192*avg
		double speed = (15625l * size) / (8192d*avg);
		System.out.printf("| %10d| %10d| %10.0f| %10.0f| %10.1f| %13.1f| \n", size/1024, sum/1000000, avg, stdDev, 100*stdDev/avg, speed == 0 ? Double.NaN : speed);
	}

	private double stdDev(long[] times, double avg, int factor) {
		double dev = 0.0;
		for (long time : times) {
			dev -= (time/(double)factor - avg) * (avg - time/(double)factor);
		}
		return Math.sqrt(dev / (times.length - 1));
	}
}

File: src/cpp/examples/pingpongcpp/Ping.cpp
/*
Copyright 2012 MAPPER consortium

GNU Lesser General Public License

This file is part of MUSCLE (Multiscale Coupling Library and Environment).

    MUSCLE is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    MUSCLE is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with MUSCLE.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include <climits>
#include <cstring>
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <sys/time.h>
#include <cppmuscle.hpp>
#include <exception>

using namespace muscle;
using namespace std;

double time_avg(const long total, const int runs, const int steps)
{
    return total/(double)(steps*runs);
}
double stddev(const long *times, const double avg, const int runs, const int steps)
{
    double dev = 0.0;
    double dsteps = steps;
    
    for (int i = 0; i < runs; i++) {
        dev += (times[i]/dsteps - avg)*(times[i]/dsteps - avg);
    }
    return sqrt(dev/(double)(runs - 1));
}

long usec_diff(struct timeval *tpStart, struct timeval *tpEnd)
{
    return (1000000l*(long)tpEnd->tv_sec + (long)tpEnd->tv_usec) -
           (1000000l*(long)tpStart->tv_sec + (long)tpStart->tv_usec);
}

void print_stats(const int printdesc, const long *times, const long totaltime, const int size_kib, const int runs, const int steps)
{
    double avg = time_avg(totaltime, runs, steps);
    double s = stddev(times, avg, runs, steps);

    if (printdesc) {
        printf("|=%10s|=%10s|=%10s|=%10s|=%10s|=%10s|\n","Size (kiB)","Total (ms)","Avg (us)","StdDev(us)","StdDev(%)","Speed (MB/s)");
    }
    double speed = size_kib*2*1024 /avg;
    printf("| %10d| %10d| %10.0f| %10.0f| %10.1f| %10.1f|\n", size_kib, (int)(totaltime/1000), avg, s, 100*s/avg, speed);
	fflush(stdout);
}

long do_computation(long* times, const int size, const int runs, const int steps)
{
    struct timeval tpStart, tpEnd;
    size_t count = size*128; // 1024/sizeof(double)
    void *buf = malloc(count*sizeof(double));
    long total = 0;

    for (int test = 0; test < runs; test++) {
        gettimeofday(&tpStart, NULL);
        
        for (int i = 0; i < steps; i++) {
            env::send("out", buf, count, MUSCLE_DOUBLE);
            env::receive("in", buf, count, MUSCLE_DOUBLE);
        }
        
        gettimeofday(&tpEnd, NULL);
        total += times[test] = usec_diff(&tpStart, &tpEnd);
    }
    
    free(buf);
    return total;
}

/**
a native kernel which sends an array of double
\author JB
*/
int main(int argc, char **argv)
{
	try
	{
		env::init(&argc, &argv);
		// How many steps for a single test
		const int steps = atoi(cxa::get_property("steps").c_str());

		// How many steps total will be done
		const int max_timesteps = atoi(cxa::get_property("max_timesteps").c_str()) + 1;
		
		// How many steps total will be done
		const int runs = atoi(cxa::get_property("same_size_runs").c_str());
		
		// size in B
		int size = atoi(cxa::get_property("start_kiB_per_message").c_str());
		
		// Number of steps to warm up the system
		const int prepare_steps = atoi(cxa::get_property("preparation_steps").c_str());

		const int tests_count = atoi(cxa::get_property("tests_count").c_str());
		
		// helper with results for a single test
		long *totalTimes = new long[runs];
		size_t preparation_size = 1024/8;
		double *data = (double *)malloc(preparation_size*sizeof(double));

		// Making noise in order to give time for JVM to stabilize
		cout << "Preparing";
		for (int i = 0; i < prepare_steps; i++)
		{
			env::send("out", data, preparation_size, MUSCLE_DOUBLE);
			env::receive("in", data, preparation_size, MUSCLE_DOUBLE);
			if (i % 5 == 0)
				cout << ".";
		}
		free(data);
		cout << "\n\nValues are NOT divided by 2. Each value is calculated for RTT." << endl;
		printf("Sending %d messages in total. For each data size, %d tests are performed, each sending %d messages.\n\n", max_timesteps, runs, steps);	
		
		for (int i = 0; i < tests_count; i++) {
			long sum = do_computation(totalTimes, size, runs, steps);
			
			print_stats(i==0, totalTimes, sum, size, runs, steps);
			if (size == 0) {
				size = 1;
			} else if (size < INT_MAX / 2) {
				size *= 2;
			} else {
				break;
			}
		}
		env::finalize();
		delete [] totalTimes;
	}
	catch(exception& ex)
	{
		logger::severe("Error occurred in Ping: %s", ex.what());
	}
	catch(...)
	{
		cerr << "Error occured in Ping" << endl;
	}
	
	return 0;
}
